Flaky Test: Race detection on MultiKueue AdmissionCheck controller

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:

The CI detects race conditions in the following:

- `sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue: TestListMultiKueueClustersUsingKubeConfig/multiple_clusters,_single_match`

```shell
{Failed  === RUN   TestListMultiKueueClustersUsingKubeConfig/multiple_clusters,_single_match
==================
WARNING: DATA RACE
Read at 0x00c000438043 by goroutine 203:
  testing.(*common).logDepth()
      /usr/local/go/src/testing/testing.go:1053 +0xca
  testing.(*common).log()
      /usr/local/go/src/testing/testing.go:1046 +0x7b
  testing.(*common).Log()
      /usr/local/go/src/testing/testing.go:1087 +0x53
  github.com/go-logr/logr/testr.logInfo()
      /home/prow/go/src/sigs.k8s.io/kueue/vendor/github.com/go-logr/logr/testr/testr.go:101 +0x193
  github.com/go-logr/logr/testr.testloggerInterface.Info()
      /home/prow/go/src/sigs.k8s.io/kueue/vendor/github.com/go-logr/logr/testr/testr.go:148 +0x13c
  github.com/go-logr/logr/testr.(*testloggerInterface).Info()
      <autogenerated>:1 +0x147
  github.com/go-logr/logr.Logger.Info()
      /home/prow/go/src/sigs.k8s.io/kueue/vendor/github.com/go-logr/logr/logr.go:280 +0x107
  sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.(*KubeConfigFSWatcher).Start.func1()
      /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue/fswatch.go:92 +0x904

Previous write at 0x00c000438043 by goroutine 197:
  testing.tRunner.func1()
      /usr/local/go/src/testing/testing.go:1779 +0x8b9
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:610 +0x5d
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1851 +0x44

Goroutine 203 (running) created at:
  sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.(*KubeConfigFSWatcher).Start()
      /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue/fswatch.go:70 +0x245
  sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.TestFSWatchAddRm.func2()
      /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue/fswatch_test.go:227 +0x44
  sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.TestFSWatchAddRm.func13()
      /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue/fswatch_test.go:452 +0x5a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1851 +0x44

Goroutine 197 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1851 +0x8f2
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:2279 +0x85
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1792 +0x225
  testing.runTests()
      /usr/local/go/src/testing/testing.go:2277 +0x96c
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:2142 +0xeea
  main.main()
      _testmain.go:71 +0x164
==================
    testing.go:1490: race detected during execution of test
--- FAIL: TestListMultiKueueClustersUsingKubeConfig/multiple_clusters,_single_match (0.04s)
}
```

- `sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue: TestListMultiKueueClustersUsingKubeConfig`

```shell
{Failed  === RUN   TestListMultiKueueClustersUsingKubeConfig
--- FAIL: TestListMultiKueueClustersUsingKubeConfig (0.13s)
}
```

**What you expected to happen**:

No race condition

**How to reproduce it (as minimally and precisely as possible)**:

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/6673/pull-kueue-test-unit-main/1960302981727391744

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/pkg/controller/admissionchecks/multikueue/fswatch_test.go
+++ b/pkg/controller/admissionchecks/multikueue/fswatch_test.go
@@ -29,6 +29,7 @@ import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
 	"k8s.io/utils/set"
+	ctrl "sigs.k8s.io/controller-runtime"
 
 	utiltesting "sigs.k8s.io/kueue/pkg/util/testing"
 )
@@ -145,8 +146,11 @@ func TestFSWatch(t *testing.T) {
 					t.Fatalf("unexpected prepare error: %s", err)
 				}
 			}
-			ctx, _ := utiltesting.ContextWithLog(t)
+			// Using an empty context here to avoid a race condition with the test context when setting the logger.
+			ctx := context.Background()
+			ctrl.LoggerInto(ctx, utiltesting.NewLogger(t))
 			ctx, cancel := context.WithTimeout(ctx, time.Second)
+			defer cancel()
 			watcher := newKubeConfigFSWatcher()
 
 			// start the recorder
@@ -196,7 +200,9 @@ func TestFSWatch(t *testing.T) {
 }
 
 func TestFSWatchAddRm(t *testing.T) {
-	ctx, _ := utiltesting.ContextWithLog(t)
+	// Using an empty context here to avoid a race condition with the test context when setting the logger.
+	ctx := context.Background()
+	ctrl.LoggerInto(ctx, utiltesting.NewLogger(t))
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 	basePath := t.TempDir()

--- a/pkg/util/testing/context.go
+++ b/pkg/util/testing/context.go
@@ -35,9 +35,13 @@ func LogLevelWithDefault(defaultLogLevel int) int {
 	return level
 }
 
-func ContextWithLog(t *testing.T) (context.Context, logr.Logger) {
-	logger := testr.NewWithOptions(t, testr.Options{
+func NewLogger(t *testing.T) logr.Logger {
+	return testr.NewWithOptions(t, testr.Options{
 		Verbosity: LogLevelWithDefault(2),
 	})
+}
+
+func ContextWithLog(t *testing.T) (context.Context, logr.Logger) {
+	logger := NewLogger(t)
 	return ctrl.LoggerInto(t.Context(), logger), logger
 }

